apps/contribution/test/contribution.js

```diff
@@ -0,0 +1,231 @@
+
+    //init token (app)
+    await token.initialize(appsId);
+
+    //create token mint permission for coin owner
+    await acl.createPermission(
+      contribution.address,
+      token.address,
+      await token.MINT_TOKEN_ROLE(),
+      root,
+      { from: root }
+    );
+
+    //init contributor app
+    await contributor.initialize(root, appsId);
+
+    await acl.createPermission(
+      root,
+      contributor.address,
+      await contributor.MANAGE_CONTRIBUTORS_ROLE(),
+      root,
+      { from: root }
+    );
+  });
+
+  describe("Owner default space permissions", async () => {
+    it('check owner can add contribution', async () => {
+      let addContributionPermission = await acl.hasPermission(root, contribution.address, await contribution.ADD_CONTRIBUTION_ROLE());
+      // eslint-disable-next-line no-undef
+      assert.equal(addContributionPermission, true);
+    });
+
+    it('check owner can veto contribution', async () => {
+      let vetoContributionPermission = await acl.hasPermission(root, contribution.address, await contribution.VETO_CONTRIBUTION_ROLE());
+      // eslint-disable-next-line no-undef
+      assert.equal(vetoContributionPermission, true);
+    });
+
+    it('check contribution app can mint token', async () => {
+      let mintTokenPermission = await acl.hasPermission(contribution.address, token.address, await token.MINT_TOKEN_ROLE());
+      // eslint-disable-next-line no-undef
+      assert.equal(mintTokenPermission, true);
+    });
+  });
+
+  describe("Add contribution", async () => {
+    let amount = 100;
+    let contributorId = 1;
+    let hashDigest = '0x0000000000000000000000000000000000000000000000000000000000000000';
+    let hashFunction = 1;
+    let hashSize = 1;
+
+    it("should revert when add contribution from address that does not have permission", async () => {
+      return assertRevert(async () => {
+        await contribution.add(amount, contributorId, hashDigest, hashFunction, hashSize, {from: member1});
+        'sender does not have permission';
+      });
+    });
+
+    it("should revert when add contribution with amount equal to zero", async () => {
+      return assertRevert(async () => {
+        await contribution.add(0, contributorId, hashDigest, hashFunction, hashSize, {from: root});
+        'amount equal to zero';
+      });
+    });
+
+    it("add contribution", async () => {
+      let contributionCountBefore = await contribution.contributionsCount();
+      await contribution.add(amount, contributorId, hashDigest, hashFunction, hashSize, {from: root});
+      let contributionCountAfter = await contribution.contributionsCount();
+      // eslint-disable-next-line no-undef
+      assert.equal(contributionCountAfter.toNumber()-contributionCountBefore.toNumber(), 1);
+      let contributionObject = await contribution.getContribution(contributionCountAfter.toNumber());
+      // eslint-disable-next-line no-undef
+      assert.equal(contributionObject[1], contributorId);
```

Comment:
I think you can also add expectation descriptions as the last argument to `assert()`, in case you really want to assert multiple things in one `it` block.

Reply:
most of those multiple assert check for the same thing, that the contribution is added.. XD

Reply:
They very line I commented on asserts that the contributorId is correct, not that the contribution itself was added.

Reply:
yeah true, honestly don't know if that all those asserts are necessary... maybe just for more code coverage

Reply:
In this case it's only important that you describe what you mean to test exactly. Currently, there's no description of what is being tested in this block.